sae_lens/training/toy_models.py

```diff
@@ -271,17 +215,141 @@ def optimize(
+            nn.init.xavier_normal_(t.empty((cfg.n_hidden, cfg.n_features)))
+        )
+        self.b_final = nn.Parameter(t.zeros((cfg.n_features)))
+        self.to(device)
+
+        # Add and setup hookpoints.
+        self.hook_hidden = HookPoint()
+        self.hook_out_prebias = HookPoint()
+        self.setup()
+
+    def forward(
+        self,
+        features: Float[Tensor, "... features"],
+        return_type: str = "reconstruction",
+    ) -> Float[Tensor, "... features"]:
+        hidden = self.hook_hidden(
+            einops.einsum(
+                features,
+                self.W,
+                "... features, hidden features -> ... hidden",
+            )
+        )
+        out = self.hook_out_prebias(
+            einops.einsum(
+                hidden,
+                self.W,
+                "... hidden, hidden features -> ... features",
+            )
+        )
+        reconstructed = F.relu(out + self.b_final)
+
+        if return_type == "loss":
+            return self.calculate_loss(reconstructed, features)
+        elif return_type == "reconstruction":
+            return reconstructed
+        else:
+            raise ValueError(f"Unknown return type: {return_type}")
+
+    def calculate_loss(
+        self,
+        out: Float[Tensor, "batch features"],
+        batch: Float[Tensor, "batch features"],
+    ) -> Float[Tensor, ""]:
+        """
+        Calculates the loss for a given batch, using this loss described in the Toy Models paper:
+
+            https://transformer-circuits.pub/2022/toy_model/index.html#demonstrating-setup-loss
+
+        Note, `model.importance` is guaranteed to broadcast with the shape of `out` and `batch`.
+        """
+        error = self.importance * ((batch - out) ** 2)
+        loss = einops.reduce(error, "batch features -> ()", "mean").sum()
+        return loss
+
+
+class ReluOutputModelCE(ReluOutputModel):
+    """
+    A variant of Anthropic's ReLU Output Model.
+    This model is trained with a Cross Entropy loss instead of MSE loss.
+    The model task is to identify which feature has the largest magnitude activation in the input.
+    The model has an extra feature dimension which is set to a constant nonzero value,
+    which allows for proper classification when all features are zero.
+    """
+
+    W: Float[Tensor, "n_hidden n_features"]
+    b_final: Tensor
+    # Our linear map is x -> ReLU(W.T @ W @ x + b_final)
+
+    def __init__(
+        self,
+        cfg: ToyConfig,
+        device: torch.device = torch.device("cpu"),
+        extra_feature_value: float = 1e-6,
+    ):
+        super().__init__(cfg)
+        self.extra_feature_value = extra_feature_value
+
+        self.W = nn.Parameter(
+            nn.init.xavier_normal_(t.empty((cfg.n_hidden, cfg.n_features + 1)))
+        )
+        self.b_final = nn.Parameter(t.zeros((cfg.n_features + 1)))
+        self.importance = _init_importance(cfg.importance, cfg.n_features + 1, device)
```

Comment:
@evanhanders I had to tweak this to get training tests to run, since otherwise `self.importance` has 1 less dim than the batch

Reply:
Ah! Good catch -- thanks for fixing this; looks right! I'm not entirely logically sure what 'importance' should be implemented as in the CE model so this seems like the right first step.